ThunderCloud/AchievementDisplayView.swift

```diff
@@ -39,83 +193,164 @@ open class AchievementDisplayView: UIView, AchievementDisplayable {
+    private func setup() {
+        backgroundColor = ThemeManager.shared.theme.backgroundColor
+        
+        updateLabels()
+        didUpdateExpirableAchievement(animated: false)
+        
+        addSubviews()
+        constrain()
+    }
     
-    override open func layoutSubviews() {
+    /// Add subviews to view hierarchy
+    private func addSubviews() {
+        expiryStackView.addArrangedSubview(expiryTitleLabel)
+        expiryStackView.addArrangedSubview(expiryDateLabel)
         
-        super.layoutSubviews()
+        progressView.addSubview(badgeImageView)
         
-        badgeImageView.center = CGPoint(x: frame.width/2, y: frame.height/2)
-        badgeImageView.backgroundColor = .white
-        badgeImageView.cornerRadius = badgeImageView.bounds.width/2
+        stackView.addArrangedSubview(titleLabel)
+        stackView.addArrangedSubview(progressView)
+        stackView.addArrangedSubview(expiryStackView)
+        stackView.addArrangedSubview(spaceView)
+        stackView.addArrangedSubview(subtitleLabel)
+        stackView.addArrangedSubview(expiryDetailLabel)
         
-        let availableSize = CGSize(width: frame.size.width - AchievementDisplayView.contentEdgePadding.left - AchievementDisplayView.contentEdgePadding.right, height: CGFloat.greatestFiniteMagnitude)
-        let titleSize = titleLabel.sizeThatFits(availableSize)
-        titleLabel.frame = CGRect(
-            x: AchievementDisplayView.contentEdgePadding.left,
-            y: badgeImageView.frame.minY - titleSize.height - AchievementDisplayView.labelImageSpacing,
-            width: frame.width - AchievementDisplayView.contentEdgePadding.left - AchievementDisplayView.contentEdgePadding.right,
-            height: titleSize.height
-        )
+        addSubview(stackView)
+    }
+      
+    /// Constrain views
+    private func constrain() {
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+        progressView.translatesAutoresizingMaskIntoConstraints = false
+        badgeImageView.translatesAutoresizingMaskIntoConstraints = false
+        spaceView.translatesAutoresizingMaskIntoConstraints = false
         
-        let size = subtitleLabel.sizeThatFits(availableSize)
-        subtitleLabel.frame = CGRect(
-            x: AchievementDisplayView.contentEdgePadding.left,
-            y: badgeImageView.frame.maxY + AchievementDisplayView.labelImageSpacing,
-            width: frame.size.width - AchievementDisplayView.contentEdgePadding.left - AchievementDisplayView.contentEdgePadding.right,
-            height: size.height
-        )
+        NSLayoutConstraint.activate([
+            // StackView
+            stackView.centerXAnchor.constraint(equalTo: centerXAnchor),
+            stackView.widthAnchor.constraint(equalTo: widthAnchor, multiplier: Constants.stackViewWidthScale),
+            
+            // BadgeView
+            progressView.widthAnchor.constraint(equalTo: stackView.widthAnchor,
```

Comment:
@simonmitchell Let's drop the parameters onto new lines when they are particularly long

Reply:
Done :D